### PR TITLE
Optimize peak heap usage

### PR DESCRIPTION
**Pull request by jcollins-g**
For testing/landing purposes, this includes #1851, but is only different from that PR by the final 3 commits, starting with: https://github.com/dart-lang/dartdoc/commit/2afd6216b7ca8bc318f769d6da7d39a4cc981ac1.

While this eliminates some extra traversal of ResolvedUnitResults and saves on peak memory, the cost is a slightly more complicated PackageGraph initialization that requires some callbacks in getLibraries.

### CODE REVIEW

**Comment by stereotype441** [comment]
@bwilkerson can you review this?

**Comment by jcollins-g** [comment]
This is now ready for review. 

https://github.com/dart-lang/dartdoc/pull/1858/files/88c67be8c85f140398df854741e75b3663e1a586..7ac903a3f0d2ec8bbf5191c0f3a31e35dd9e33ac

File: lib/src/model.dart [hunk 1]
```diff
@@ -6652,7 +6722,7 @@ class PackageBuilder {
         }
       }
     } while (!lastPass.containsAll(current));
-    return libraries;
+    return;
```
Comment (bwilkerson) [inline lib/src/model.dart]: style nit: consider removing this line

File: lib/src/model_utils.dart [hunk 2]
```diff
@@ -7,6 +7,7 @@ library dartdoc.model_utils;
 import 'dart:convert';
 import 'dart:io';
 
+import 'package:analyzer/analyzer.dart';
```
Comment (bwilkerson) [inline lib/src/model_utils.dart]: Please also replace this import.

File: lib/src/model.dart [hunk 3]
```diff
@@ -10,6 +10,7 @@ import 'dart:collection' show UnmodifiableListView;
 import 'dart:convert';
 import 'dart:io';
 
+import 'package:analyzer/analyzer.dart';
```
[the remark this answers is not in the document]
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart [hunk 4]
```diff
@@ -2060,8 +2065,41 @@ abstract class GetterSetterCombo implements ModelElement {
   Accessor get setter;
 }
 
+/// Find all hashable children of a given element that are defined in the
+/// [LibraryElement] given at initialization.
+class _HashableChildLibraryElementVisitor
+    extends GeneralizingElementVisitor<void> {
+  final void Function(Element) libraryProcessor;
+  _HashableChildLibraryElementVisitor(this.libraryProcessor) {}
```
[the remark this answers is not in the document]
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model_utils.dart (hunk 2 above)
Comment (jcollins-g) [reply lib/src/model_utils.dart]: done

**Comment by jcollins-g** [comment]
Going to land this by force-landing #1851 and then doing the git dance to land this one.

**Pull request merged**